feat(curriculum): add Python error handling review page (#60559)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/review-error-handling/67f39bcb48373420f4f12d62.md

- **TypeError**: This is the error Python throws when you perform an operation on two or more incompatible data types. For example, if you try to add a string to a number, you'll get the error `TypeError: can only concatenate str (not "int") to str`.
- **IndexError**: You'll get an `IndexError` if you access an index that does not exist in a list or other sequences like tuple and string. For example, in a `Hello world` string, the index of the last character is `11`. If you go ahead and access a character this way, `greet = "hello world"; print(greet[12])`, you'll get an error with the message `IndexError: string index out of range`.
- **AttributeError**: Python raises this error when you try to use a method or property that does not exist in an object of that type. For example, calling `.append()` on a string like `"hello".append("!")` will lead to an error with the message `AttributeError: 'str' object has no attribute 'append'`.

## Good Debugging Techniques in Python

> - **Using Print Statements**: Inserting `print()` statements around various points in your code while debugging helps you see the values of variables and how your code flows.

Reviewer: ```suggestion
- **Using the `print` function**: Inserting `print()` statements around various points in your code while debugging helps you see the values of variables and how your code flows.
```

- **Using Python's Built-in Debugger (`pdb`)**: Python provides a `pdb` module for debugging. It's a part of the Python's standard library, so it's always available to use. With `pdb`, you can set a trace with the `set_trace()` method so you can start stepping through the code and inspect variables in an interactive way.
- **Leveraging IDE Debugging Tools**: Many integrated development environments (IDEs) and code editors like Pycharm and VS Code offer debugging tools with breakpoints, step execution, variable inspection, and other debugging features.
- **Writing Unit Tests**: Writing unit tests for your Python code lets you catch bugs before they become an issue and make your code behave as expected. Python provides a `unittest` module through its standard library with which you can write unit tests.

> ## How Exception Handling Works in Python

Reviewer: ```suggestion
## Exception Handling
```

- **`try...except`**: This is used to execute a block of code that might raise an exception. The `try` block is  where you anticipate an error might occur, while the `except` block takes a specified exception and runs if that specified error is raised. Here's an example:

  ```py
  try:
    print(22 / 0)
  except ZeroDivisionError:
    print('You can\'t divide by zero!')
    # You can't divide by zero!
  ```

  You can also chain multiple `except` blocks so you can handle more types of exceptions:

  ```py
  try:
    number = int(input('Enter a number: '))
    print(22 / number)
  except ZeroDivisionError:

>     print('You can\'t divide by zero!')

Reviewer: ```suggestion
    print('You cannot divide by zero!')
```

>     # You can't divide by zero! prints when you enter 0

Reviewer: ```suggestion
    # You cannot divide by zero! prints when you enter 0
```

  ```

- **`else` and `finally`**: These blocks extend `try...except`. If no exception occurs, the `else` block runs. The `finally` block always runs regardless of errors.

  ```py
  try:
    result = 100 / 4
  except ZeroDivisionError:
    print('You cannot divide by zero!') # This will not run
  else:
    print(f'Result is {result}') # Result is 25.0
  finally:
    print('Execution complete!') # Execution complete!
  ```

- **Exception Object**: This lets you access the exception itself for better debugging and printing the direct error message. To access the exception object, you need to use the `as` keyword. Here's an example:

  ```py
  try:
      value = int('This will raise an error')
  except ValueError as e:
      print(f'Caught an error: {e}')
      # Caught an error: invalid literal for int() with base 10: 'This will raise an error'
  ```

- **The `raise` Statement**: This allows you to manually raise an exception. You can use it to throw an exception when a certain condition is met. Here's an example:

  ```py
  def divide(a, b):
      if b == 0:
          raise ZeroDivisionError('You cannot divide by zero')
      return a / b
  ```

>   

Reviewer: I would add another header for this part
```suggestion
  
## Exception Signaling
  
```

- **NameError**: Python raises a `NameError` when you try to access a variable or function you have not defined. For instance, if you have the line `print(username)` in your code without having a `username` variable defined first, you will get a name error with the message `NameError: name 'username' is not defined`.
- **TypeError**: This is the error Python throws when you perform an operation on two or more incompatible data types. For example, if you try to add a string to a number, you'll get the error `TypeError: can only concatenate str (not "int") to str`.
- **IndexError**: You'll get in `IndexError` if you access an index that does not exist in a list or other sequences like tuple and string. For example, in an `Hello world` string, the index of the last character is `11`. If you go ahead and access a character this way `greet = "hello world"; print(greet[12])`, you'll get an error with the message `IndexError: string index out of range`.

> - **AttributeError**: Python raises this error when you try to use a method or property that does not exist in an object of that type. For example, calling `.append()` on a string like `"hello".append("!")` will lead to an error wiuth the message `AttributeError: 'str' object has no attribute 'append'`.

Reviewer 2: ```suggestion
- **AttributeError**: Python raises this error when you try to use a method or property that does not exist in an object of that type. For example, calling `.append()` on a string like `"hello".append("!")` will lead to an error with the message `AttributeError: 'str' object has no attribute 'append'`.
```

## Good Debugging Techniques in Python

> - **Using Print Statements**: Inserting `print()` statements around various points in your code while debuggin helps you see the values of variables and how your code flows.

Reviewer 2: ```suggestion
- **Using Print Statements**: Inserting `print()` statements around various points in your code while debugging helps you see the values of variables and how your code flows.
```

> - **Using Python's Built-in Debugger (`pdb`)**: Python provides a `pdb` module for debugging. It's a part of the Python's standard library, so it's always available to use. With `pdb`, you can set a trace with the `set_trace()` method so you can start stepping through the code and inpect variables in an interractive way.

Reviewer 2: ```suggestion
- **Using Python's Built-in Debugger (`pdb`)**: Python provides a `pdb` module for debugging. It's a part of the Python's standard library, so it's always available to use. With `pdb`, you can set a trace with the `set_trace()` method so you can start stepping through the code and inspect variables in an interactive way.
```

- **Leveraging IDE Debugging Tools**: Many integrated development environments (IDEs) and code editors like Pycharm and VS Code offer debugging tools with breakpoints, step execution, varaible inspection, and other debugging features.
- **By Writing Unit Tests**: Writing unit tests for your Python code lets you catch bugs before they become an issue and make your code behave as expected. Python provides a `unittest` module through its standard library with which you can write unit tests.

## How Exception Handling Work in Python

> - **`try...except`** lets you executable a code against a certain exception. The `try` block is the block of code where you anticipate an error might occur, while the `except` block takes a specified exception and runs if that specified error is raised. Here's an example:

Reviewer 2: ```suggestion
- **`try...except`** allows you to run a block of code and handle specific exceptions that might occur. The `try` block is the block of code where you anticipate an error might occur, while the `except` block takes a specified exception and runs if that specified error is raised. Here's an example:
```

  ```py
  try:
    print(22 / 0)
  except ZeroDivisionError:
    print('You can\'t divide by zero!')
    # You can't divide by zero!
  ```

  You can also chain multiple `except` blocks so you can handle more types of exceptions:

  ```py
  try:
    number = int(input('Enter a number: '))
    print(22 / number)
  except ZeroDivisionError:

>     print('You can\'t divide by zero')

Reviewer 2: Very small inconsistency with other error messages :)

```suggestion
    print('You can\'t divide by zero!')
```
